cloudsmith/resource_entitlement.go

```diff
@@ -0,0 +1,242 @@
+	d.Set("limit_date_range_from", entitlement.LimitDateRangeFrom)
+	d.Set("limit_date_range_to", entitlement.LimitDateRangeTo)
+	d.Set("limit_num_clients", entitlement.LimitNumClients)
+	d.Set("limit_num_downloads", entitlement.LimitNumDownloads)
+	d.Set("limit_package_query", entitlement.LimitPackageQuery)
+	d.Set("limit_path_query", entitlement.LimitPathQuery)
+	d.Set("name", entitlement.Name)
+	d.Set("token", entitlement.Token)
+
+	// namespace and repository are not returned from the entitlement read
+	// endpoint, so we can use the values stored in resource state. We rely on
+	// ForceNew to ensure if either changes a new resource is created.
+	d.Set("namespace", namespace)
+	d.Set("repository", repository)
+
+	return nil
+}
+
+func resourceEntitlementUpdate(d *schema.ResourceData, m interface{}) error {
+	pc := m.(*providerConfig)
+
+	isActive := getIsActiveFromResource(d)
+	limitDateRangeFrom := d.Get("limit_date_range_from").(string)
+	limitDateRangeTo := d.Get("limit_date_range_to").(string)
+	limitNumClients := int64(d.Get("limit_num_clients").(int))
+	limitNumDownloads := int64(d.Get("limit_num_downloads").(int))
+	limitPackageQuery := d.Get("limit_package_query").(string)
+	limitPathQuery := d.Get("limit_path_query").(string)
+	namespace := d.Get("namespace").(string)
+	repository := d.Get("repository").(string)
+	name := d.Get("name").(string)
+	token := d.Get("token").(string)
+
+	opts := &cloudsmith.EntitlementsPartialUpdateOpts{
+		Data: optional.NewInterface(cloudsmith.EntitlementsPartialUpdate{
+			IsActive:           isActive,
+			LimitDateRangeFrom: limitDateRangeFrom,
+			LimitDateRangeTo:   limitDateRangeTo,
+			LimitNumClients:    limitNumClients,
+			LimitNumDownloads:  limitNumDownloads,
+			LimitPackageQuery:  limitPackageQuery,
+			LimitPathQuery:     limitPathQuery,
+			Name:               name,
+			Token:              token,
+		}),
+		ShowTokens: optional.NewBool(true),
+	}
+
+	entitlement, _, err := pc.APIClient.EntitlementsApi.EntitlementsPartialUpdate(pc.Auth, namespace, repository, d.Id(), opts)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(entitlement.SlugPerm)
+
+	return resourceEntitlementRead(d, m)
+}
+
+func resourceEntitlementDelete(d *schema.ResourceData, m interface{}) error {
+	pc := m.(*providerConfig)
+
+	namespace := d.Get("namespace").(string)
+	repository := d.Get("repository").(string)
+
+	_, err := pc.APIClient.EntitlementsApi.EntitlementsDelete(pc.Auth, namespace, repository, d.Id())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//nolint:funlen
+func resourceEntitlement() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceEntitlementCreate,
+		Read:   resourceEntitlementRead,
+		Update: resourceEntitlementUpdate,
+		Delete: resourceEntitlementDelete,
+
+		Schema: map[string]*schema.Schema{
+			"is_active": {
+				Type:        schema.TypeBool,
+				Description: "If enabled, the token will allow downloads based on configured restrictions (if any).",
```

Comment:
Is there any way to derive these from bindings?

Reply:
They're only included in comments on the struct fields. We can read them but requires some AST parsing, a good future improvement perhaps.

Reply:
A better way is probably to generate a strings table of sorts based on the OpenAPI spec, then use that in as part of the program. Future problem.